File: sdk/go/pack/pack.go
// Package pack is the Go SDK surface for the memQL pack browser -- the
// read-only enumeration of a node's embedded + plugin-registered .memql
// tree. It powers the cockpit's "browse packs" view: list the DSL domains
// a node carries, list the .memql files inside a domain, and fetch a
// single file's source.
//
// The package mirrors sdk/go/sense in shape -- a Client that wraps a
// client.Dispatcher, typed methods that take and return SDK-owned values
// (not raw protobuf), and errors that bubble up instead of being silently
// swallowed.
//
// There is no write path: a pack subtree is owned by its registering Go
// module and the embedded tree is immutable. This surface is read-only by
// construction.
//
// Types are SDK-owned. No memqlv1.* import leaks out of this package's
// exported surface -- callers depend on pack.Domain / pack.File / etc.,
// not on protobuf.
//
// See sdk/go/CLAUDE.md for the broader SDK rules (named-primitive
// contract, opaque types, generated-code-is-read-only).
package pack

import (
	"context"
	"fmt"

	memqlv1 "github.com/znasllc-io/memql/component/grpc/gen"

	"github.com/znasllc-io/memql/sdk/go/client"
)

// Client exposes the three read-only pack-browsing operations over a
// Dispatcher. Safe to reuse across goroutines once constructed; the
// underlying Dispatcher is the multiplex point.
type Client struct {
	dispatcher *client.Dispatcher
}

// NewClient wires a pack client to the supplied dispatcher. The
// dispatcher must already be Run(); Client.* methods just dispatch
// SendAndWait calls through it.
func NewClient(dispatcher *client.Dispatcher) *Client {
	return &Client{dispatcher: dispatcher}
}

// Domain is one browsable DSL namespace on a node. Origin is "embedded"
// for a core baked-in domain or "pack:<name>" for a domain contributed at
// init() time by an external Go module (a "pack"). FileCount is the
// recursive count of browsable files (.memql + .tmpl) under the domain.
type Domain struct {
	Name      string
	Origin    string
	FileCount int
}

// File is one browsable file inside a domain. Path is relative to the
// domain root (e.g. "queries.memql", "prompts/agentReply.tmpl"); Size is
// the byte length.
type File struct {
	Path string
	Size int64
}

// FileSource is the result of ReadFile -- a file's source plus its origin
// label. Origin is "embedded" | "pack:<domain>" | "disk:<path>" (the last
// when a MEMQL_DSL_PATH override is active for the construct type). Found
// is false when the requested file does not exist; Source is then empty.
type FileSource struct {
	Domain string
	Path   string
	Source string
	Origin string
	Found  bool
}

// ListDomains returns every browsable DSL domain the node carries --
// embedded core domains plus plugin-registered pack domains -- sorted by
// name.
func (c *Client) ListDomains(ctx context.Context) ([]Domain, error) {
	msg := &memqlv1.MemqlClientMessage{
		Payload: &memqlv1.MemqlClientMessage_ListPackDomains{
			ListPackDomains: &memqlv1.ListPackDomainsMsg{},
		},
	}
	resp, err := c.dispatcher.SendAndWait(ctx, msg)
	if err != nil {
		return nil, fmt.Errorf("pack.listDomains: %w", err)
	}
	result := resp.GetListPackDomainsResult()
	if result == nil {
		return nil, nil
	}
	out := make([]Domain, 0, len(result.GetDomains()))
	for _, d := range result.GetDomains() {
		out = append(out, Domain{
			Name:      d.GetName(),
			Origin:    d.GetOrigin(),
			FileCount: int(d.GetFileCount()),
		})
	}
	return out, nil
}

// ListFiles returns the browsable files under a single domain, sorted by
// relative path. An unknown domain yields an empty slice (not an error);
// callers distinguish "no such domain" via ListDomains.
func (c *Client) ListFiles(ctx context.Context, domain string) ([]File, error) {
	msg := &memqlv1.MemqlClientMessage{
		Payload: &memqlv1.MemqlClientMessage_ListPackFiles{
			ListPackFiles: &memqlv1.ListPackFilesMsg{Domain: domain},
		},
	}
	resp, err := c.dispatcher.SendAndWait(ctx, msg)
	if err != nil {
		return nil, fmt.Errorf("pack.listFiles: %w", err)
	}
	result := resp.GetListPackFilesResult()
	if result == nil {
		return nil, nil
	}
	out := make([]File, 0, len(result.GetFiles()))
	for _, f := range result.GetFiles() {
		out = append(out, File{
			Path: f.GetPath(),
			Size: f.GetSize(),
		})
	}
	return out, nil
}

// ReadFile fetches a single file's source. relPath is relative to the
// domain root (as returned by ListFiles). The Go error return is reserved
// for wire-level failures; a missing file is reported as a FileSource with
// Found=false (not an error), mirroring the engine's read-only semantics.
func (c *Client) ReadFile(ctx context.Context, domain, relPath string) (*FileSource, error) {
	msg := &memqlv1.MemqlClientMessage{
		Payload: &memqlv1.MemqlClientMessage_ReadPackFile{
			ReadPackFile: &memqlv1.ReadPackFileMsg{
				Domain: domain,
				Path:   relPath,
			},
		},
	}
	resp, err := c.dispatcher.SendAndWait(ctx, msg)
	if err != nil {
		return nil, fmt.Errorf("pack.readFile: %w", err)
	}
	result := resp.GetReadPackFileResult()
	if result == nil {
		return nil, nil
	}
	return &FileSource{
		Domain: result.GetDomain(),
		Path:   result.GetPath(),
		Source: result.GetSource(),
		Origin: result.GetOrigin(),
		Found:  result.GetFound(),
	}, nil
}

File: sdk/ts/src/pack/pack.ts
// The pack browser: what .memql files a cluster carries, and their source.
//
// Three single-reply calls over the dispatcher, in the shape of ConstructsClient.
// The Go SDK's sdk/go/pack is the mirror; the names are kept parallel on purpose.
//
// A MISSING FILE IS NOT AN ERROR. The engine answers ReadPackFile with
// found=false and no wire error, so readFile resolves with found=false; only
// a queryError or an unrecognised envelope throws. A caller rendering a file
// that is not there should say so, not show an exception.

import type { Dispatcher } from "../client/dispatcher.js";
import { newShortId } from "../client/id.js";
import { readServerPayload } from "../client/wire.js";
import type { PackDomainWire, PackFileWire } from "../client/wire.js";

export interface PackDomain {
  name: string;
  /** "embedded" | "pack:<domain>" */
  origin: string;
  fileCount: number;
}

export interface PackFile {
  /** Relative to the domain root, e.g. "queries.memql" or "prompts/x.tmpl". */
  path: string;
  size: number;
}

export interface PackFileSource {
  domain: string;
  path: string;
  source: string;
  /** "embedded" | "pack:<domain>" | "disk:<path>" */
  origin: string;
  /** false when the file does not exist -- a normal answer, not an error. */
  found: boolean;
}

export interface PackCallOptions {
  signal?: AbortSignal;
}

export class PackClient {
  constructor(private readonly dispatcher: Dispatcher) {
    if (!dispatcher) throw new Error("PackClient: dispatcher is required");
  }

  async listDomains(opts: PackCallOptions = {}): Promise<PackDomain[]> {
    const reply = await this.dispatcher.sendAndWait(
      { listPackDomains: { requestId: newShortId() } },
      opts.signal,
    );
    const payload = readServerPayload(reply);
    if (payload?.kind === "queryError") {
      throw new Error(`listDomains: ${payload.value.error?.message ?? "(no message)"}`);
    }
    if (payload?.kind !== "listPackDomainsResult") {
      throw new Error("listDomains: unexpected reply envelope");
    }
    return (payload.value.domains ?? []).map(domainFromWire);
  }

  async listFiles(domain: string, opts: PackCallOptions = {}): Promise<PackFile[]> {
    const reply = await this.dispatcher.sendAndWait(
      { listPackFiles: { requestId: newShortId(), domain } },
      opts.signal,
    );
    const payload = readServerPayload(reply);
    if (payload?.kind === "queryError") {
      throw new Error(`listFiles: ${payload.value.error?.message ?? "(no message)"}`);
    }
    if (payload?.kind !== "listPackFilesResult") {
      throw new Error("listFiles: unexpected reply envelope");
    }
    return (payload.value.files ?? []).map(fileFromWire);
  }

  async readFile(domain: string, path: string, opts: PackCallOptions = {}): Promise<PackFileSource> {
    const reply = await this.dispatcher.sendAndWait(
      { readPackFile: { requestId: newShortId(), domain, path } },
      opts.signal,
    );
    const payload = readServerPayload(reply);
    if (payload?.kind === "queryError") {
      throw new Error(`readFile: ${payload.value.error?.message ?? "(no message)"}`);
    }
    if (payload?.kind !== "readPackFileResult") {
      throw new Error("readFile: unexpected reply envelope");
    }
    const v = payload.value;
    return {
      domain: v.domain ?? domain,
      path: v.path ?? path,
      source: v.source ?? "",
      origin: v.origin ?? "",
      found: v.found === true,
    };
  }
}

// Every field is defaulted: protojson omits zero values, so a raw read hands
// the consumer `undefined` for exactly the common case.
function domainFromWire(d: PackDomainWire): PackDomain {
  return { name: d.name ?? "", origin: d.origin ?? "", fileCount: d.fileCount ?? 0 };
}

function fileFromWire(f: PackFileWire): PackFile {
  const raw = f.size;
  const size = typeof raw === "string" ? Number(raw) : (raw ?? 0);
  return { path: f.path ?? "", size: Number.isFinite(size) ? size : 0 };
}
